### PR TITLE
[MODEL] update hybrid_forward to use deferred compute

### PR DESCRIPTION
**Pull request by szha**
## Description ##
update hybrid_forward to use deferred compute

### Essentials ###
- [x] PR's title starts with a category (e.g. [BUGFIX], [MODEL], [TUTORIAL], [FEATURE], [DOC], etc)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented

### Changes ###
- [x] update hybrid_forward to use deferred compute

cc @dmlc/gluon-nlp-team

### CODE REVIEW

**Comment by szha** [comment]
At the moment this is still blocked by https://github.com/apache/incubator-mxnet/issues/19286

**Comment by sxjscience** [comment]
What's the status of this PR?

**Comment by szha** [comment]
blocked on the issue of GPT2 as communicated to you prior offline..

**Comment by sxjscience** [comment]
Are there issues other than the GPT-2 ?

**Comment by szha** [comment]
no. all other checks have passed.

**Comment by leezu** [comment]
GPT2 test case to be unblocked by https://github.com/apache/incubator-mxnet/pull/19374

**Comment by szha** [comment]
looks like it's still failing for the same issue

**Comment by leezu** [comment]
> looks like it's still failing for the same issue

As there are no new binaries since a few days

**Comment by leezu** [comment]
@barry-jin can you help update the docker container to include the most recent mxnet build? In the mid-term the docker-container should not include a fixed version of mxnet but we should install the latest nightly build upon every CI run (similar to how the prior setup would run the conda env step during each job: https://github.com/dmlc/gluon-nlp/blob/e65cd41e872fb174037702591ec86d7c9ba62b60/ci/batch/docker/gluon_nlp_job.sh#L23-L24)

**Comment by leezu** [comment]
It appears the GPU tests still run with an outdated version of MXNet.

I noticed that the MXNet version is fixed in the container:

https://github.com/dmlc/gluon-nlp/blob/46c9d014ac0a022b208ce335cb6ee8a54771bae4/tools/docker/ubuntu18.04-gpu.Dockerfile#L35

Though @barry-jin helped point out that it should later be updated at:

https://github.com/dmlc/gluon-nlp/blob/46c9d014ac0a022b208ce335cb6ee8a54771bae4/tools/docker/gluon_nlp_job.sh#L25-L34

**Comment by sxjscience** [comment]
Should we add `--user` to the gluon_nlp_job.sh?

File: src/gluonnlp/layers.py
Comment (sxjscience) [inline src/gluonnlp/layers.py:28]: For me I feel `onp` might be better, although `_np` is not unacceptable.

**Comment by sxjscience** [comment]
LGTM overall, we may try to merge after the pytest has passed.

**Review by leezu — APPROVED** [review]

**Comment by szha** [comment]
@leezu thanks for the help in unblocking this PR.

**Pull request merged**
